k8s/envoy-opa/README.md

```diff
@@ -0,0 +1,357 @@
+```
+One thing to note is the use of the `openpolicyagent/opa:latest-istio` image. This image is created by retagging the latest released image of OPA-Istio which extends OPA with a gRPC server that implements the Envoy External Authorization API so OPA can communicate policy decisions with Envoy.
+
+The ConfigMap `backend-opa-policy` needs to be added into the `volumeMounts` section, like this:
+
+```console
+- name: backend-opa-policy
+   configMap:
+      name: backend-opa-policy-config
+```
+
+The ConfigMap `backend-opa-policy` provides two resources, `opa-config.yaml` described in [OPA Configuration](#opa-configuration) and the `opa-policy.rego` policy explained in the [Rego Policy](#opa-policy) section.
+
+## OPA Configuration
+
+For this tutorial we create the following OPA configuration file in [`opa-config.yaml`](./k8s/backend/config/opa-config.yaml):
+
+```console
+decision_logs:
+   console: true
+plugins:
+   envoy_ext_authz_grpc:
+      addr: :8182
+      query: data.envoy.authz.allow
+```
+
+Here, `decision_logs.console: true` forces OPA to log the decisions locally at info level to the console. Later on in the tutorial we'll use these logs to examine the results for different requests.
+
+Next, let's review the configuration for the `envoy_ext_authz_grpc` plugin. First, the `addr` key sets the listening address of the Envoy External Authorization gRPC server. This must match the value configured in the Envoy Filter resource detailed in a following section.
+The `query` key defines the name of the policy to query. The next section focuses on details of the `envoy.authz.allow` policy specified for the `query` key.
+
+## OPA Policy
+
+OPA policies are expressed in a high-level declarative language called Rego. For this tutorial we created a sample rule named `allow` that includes three expressions (see [`opa-policy.rego`](./k8s/backend/config/opa-policy.rego)). All the expressions must evaluate to true for the rule to be true.
+
+```console
+default allow = false
+
+allow {
+   valid_path
+   http_request.method == "GET"
+   svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend"
+}
+```
+
+Let's take a look at each expression individually. `valid_path` is a user-defined function created to ensure that only requests sent to permitted resources are allowed.
+
+```console
+import input.attributes.request.http as http_request
+
+valid_path {
+   glob.match("/balances/*", [], http_request.path)
+}
+
+valid_path {
+   glob.match("/profiles/*", [], http_request.path)
+}
+
+valid_path {
+   glob.match("/transactions/*", [], http_request.path)
+}
+```
+
+The function `valid_path` makes use of the built-in function `glob.match(`_pattern_`, `_delimiters_`, `_match_`)` the output of which is true if _match_ can be found in _pattern_ which is separated by _delimiters_. Then, to express logical OR in Rego you define multiple rules with the same name. That is why there are three definitions of `valid_path`, one per valid resource.
```

Comment:
The commas in monospace look funny because they are gray.

```suggestion
The function `valid_path` makes use of the built-in function `glob.match(`_pattern, delimiters, match_`)` the output of which is true if _match_ can be found in _pattern_ which is separated by _delimiters_. Then, to express logical OR in Rego you define multiple rules with the same name. That is why there are three definitions of `valid_path`, one per valid resource.
```